src/libexpr/primops.cc

```diff
@@ -54,6 +54,7 @@ void EvalState::realiseContext(const PathSet & context)
         auto ctx = store->parseStorePath(decoded.first);
         if (!store->isValidPath(ctx))
             throw InvalidPathError(store->printStorePath(ctx));
+        realisedPaths.push_back(ctx.clone());
         if (!decoded.second.empty() && ctx.isDerivation()) {
             drvs.push_back(StorePathWithOutputs{ctx.clone(), {decoded.second}});
 
```

Comment:
Err... don't we need to do something like this?
```suggestion
        if (decoded.second.empty()) {
            realisedPaths.push_back(ctx.clone());
        } else {
            assert (ctx.isDerivation());
            drvs.push_back(ctx.clone(), {decoded.second}})
            // ... (can't include in suggestion)
            realisedPaths.push_back(i->second.path);
```

Reply:
i'm unsure what you mean, but I think in this case just the part before the `!` is fine. computeFSClosure doesn't care about outputs.

Reply:
Ah! I forgot we were requiring `recursive`

Reply:
Ah! I forgot we were requiring `recursive`

Reply:
or it doesn't compute dependencies of outputs at least

Reply:
I can't marked resolved but I'm satisfied as to all these.